File: android/motion-gesture-recorder/src/main/kotlin/io/github/mtatsuto/motiongesture/recorder/MotionTraceOutput.kt
package io.github.mtatsuto.motiongesture.recorder

import java.nio.ByteBuffer
import java.nio.channels.FileChannel
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption
import java.nio.file.StandardOpenOption
import java.util.UUID

enum class MotionTraceWriteDisposition {
    WRITTEN,
    BACKPRESSURED,
}

interface MotionTraceOutput {
    val temporaryPath: Path?
    val destinationPath: Path?

    fun start()
    fun write(data: ByteArray): MotionTraceWriteDisposition
    fun commit(): Path?
    fun abortPreservingPartial()
}

/** Streams to a same-directory temporary file and commits with an atomic move. */
class AtomicFileMotionTraceOutput(
    override val destinationPath: Path,
) : MotionTraceOutput {
    override var temporaryPath: Path? = null
        private set

    private var channel: FileChannel? = null
    private var committed = false

    override fun start() {
        check(channel == null && !committed) { "output is not idle" }
        require(destinationPath.fileName.toString().endsWith(".mge.jsonl")) {
            "destination must end in .mge.jsonl"
        }
        require(Files.notExists(destinationPath)) { "destination already exists" }
        val parent = requireNotNull(destinationPath.toAbsolutePath().parent)
        require(Files.isDirectory(parent)) { "destination parent directory does not exist" }
        val temporary = Files.createTempFile(
            parent,
            ".${destinationPath.fileName}.${UUID.randomUUID()}.",
            ".partial",
        )
        temporaryPath = temporary
        channel = FileChannel.open(temporary, StandardOpenOption.WRITE)
    }

    override fun write(data: ByteArray): MotionTraceWriteDisposition {
        val output = checkNotNull(channel) { "output is not started" }
        val buffer = ByteBuffer.wrap(data)
        while (buffer.hasRemaining()) output.write(buffer)
        return MotionTraceWriteDisposition.WRITTEN
    }

    override fun commit(): Path? {
        val output = checkNotNull(channel) { "output is not started" }
        val temporary = checkNotNull(temporaryPath)
        output.force(true)
        output.close()
        channel = null
        Files.move(temporary, destinationPath, StandardCopyOption.ATOMIC_MOVE)
        committed = true
        return destinationPath
    }

    override fun abortPreservingPartial() {
        runCatching { channel?.force(true) }
        runCatching { channel?.close() }
        channel = null
    }
}

File: swift/Sources/MotionGestureRecorder/MotionTraceOutput.swift
import Foundation

public enum MotionTraceWriteDisposition: Equatable, Sendable {
  case written
  case backpressured
}

/// Injected byte destination used by the recorder for deterministic tests and local files.
///
/// Implementations must not expose a destination as committed before `commit()` succeeds.
public protocol MotionTraceOutput: AnyObject {
  var temporaryURL: URL? { get }
  var destinationURL: URL? { get }

  func start() throws
  func write(_ data: Data) throws -> MotionTraceWriteDisposition
  func commit() throws -> URL?
  func abortPreservingPartial()
}

/// Streams an uncompressed trace to a same-directory temporary file and atomically renames it.
public final class AtomicFileMotionTraceOutput: MotionTraceOutput {
  public let destinationURL: URL?
  public private(set) var temporaryURL: URL?

  private let fileManager: FileManager
  private var handle: FileHandle?
  private var committed = false

  public init(destinationURL: URL, fileManager: FileManager = .default) {
    self.destinationURL = destinationURL
    self.fileManager = fileManager
  }

  public func start() throws {
    guard handle == nil, !committed, let destinationURL else {
      throw AtomicFileOutputError.invalidState
    }
    guard destinationURL.pathExtension == "jsonl",
      destinationURL.lastPathComponent.hasSuffix(".mge.jsonl")
    else {
      throw AtomicFileOutputError.invalidSuffix
    }
    guard !fileManager.fileExists(atPath: destinationURL.path) else {
      throw AtomicFileOutputError.destinationExists
    }

    let parent = destinationURL.deletingLastPathComponent()
    var isDirectory: ObjCBool = false
    guard fileManager.fileExists(atPath: parent.path, isDirectory: &isDirectory),
      isDirectory.boolValue
    else {
      throw AtomicFileOutputError.parentDirectoryMissing
    }

    let temporaryURL = parent.appendingPathComponent(
      ".\(destinationURL.lastPathComponent).\(UUID().uuidString.lowercased()).partial"
    )
    guard fileManager.createFile(atPath: temporaryURL.path, contents: nil) else {
      throw AtomicFileOutputError.cannotCreateTemporaryFile
    }
    self.temporaryURL = temporaryURL
    handle = try FileHandle(forWritingTo: temporaryURL)
  }

  public func write(_ data: Data) throws -> MotionTraceWriteDisposition {
    guard let handle, !committed else { throw AtomicFileOutputError.invalidState }
    try handle.write(contentsOf: data)
    return .written
  }

  public func commit() throws -> URL? {
    guard let handle, let temporaryURL, let destinationURL, !committed else {
      throw AtomicFileOutputError.invalidState
    }
    try handle.synchronize()
    try handle.close()
    self.handle = nil
    try fileManager.moveItem(at: temporaryURL, to: destinationURL)
    committed = true
    return destinationURL
  }

  public func abortPreservingPartial() {
    try? handle?.synchronize()
    try? handle?.close()
    handle = nil
  }
}

enum AtomicFileOutputError: Error {
  case invalidState
  case invalidSuffix
  case destinationExists
  case parentDirectoryMissing
  case cannotCreateTemporaryFile
}
